Efficient index creation

joka921 opened the pull request:
Efficient index creation for RDF-NTriples

niklas88 commented:
Ok @joka921 I've made the Travis build use GCC 5, can you try a `gir rebase --interactive upstream master`, dropping the duplicated commits and squashing the fix up commits into the ones with real changes?

niklas88 commented:
looks like you stil merged instead of rebased?

niklas88 commented:
@joka921 sorry I was wrong, apparently rebase doesn't work across remotes so you really do need the upstream tracking branch.

niklas88 commented:
So I tried this with your changes on my version and here is what I did:

   git checkout -b efficient-index-creation
   git pull git://github.com/joka921/QLever.git f.efficientIndexCreation
   git rebase --interactive master # my master is the same as this version so updated with --ff-only

Then a vim window opens and I replaced pick with drop for the duplicated old commits 
 (it shows instructions at the bottom). And fixup for the commits that are just tiny fix ups like removing the unused variable. And then you need to `git push --force` to your github

niklas88 commented on src/index/ExternalVocabulary.cpp:
```diff
@@ -3,6 +3,9 @@
 
 #include "./ExternalVocabulary.h"
+
+#include <fstream>
```
System headers go before custom headers

joka921 commented:
@niklas88  Thank you for your help. After hardresetting, rebasing and force-pushing I think I got a single-commit version online which I consider to be on the state I actually wanted it. Thank you for your help, I think I have learned something for the future.

niklas88 commented on src/index/Index.cpp:
```diff
@@ -210,43 +234,71 @@ void Index::passTsvFileIntoIdVector(const string& tsvFile, ExtVec& data,
     if (ad_utility::isXsdValue(spo[2])) {
       spo[2] = ad_utility::convertValueLiteralToIndexWord(spo[2]);
-    }
+    }    
```
superfluous spaces before end of line

niklas88 commented on src/index/Index.cpp:
```diff
@@ -210,43 +234,71 @@ void Index::passTsvFileIntoIdVector(const string& tsvFile, ExtVec& data,
+
     ++i;
     if (i % 10000000 == 0) {
+      std::cout << "Lines processed: " << i << '\n';
```
Are these always direct `std::cout` instead of `LOG(INFO)`, I'm not sure at the moment?

niklas88 commented on src/index/Index.cpp:
```diff
@@ -210,43 +234,71 @@ void Index::passTsvFileIntoIdVector(const string& tsvFile, ExtVec& data,
+
+    if (i % linesPerPartial == 0) {
       LOG(INFO) << "Lines processed: " << i << '\n';
+      std::cout << "writing partial vocab no. " << numFiles << std::endl;
```
This should definitely be `LOG(INFO)` and move it to the point where we are actually writing the file

niklas88 commented on src/index/Index.cpp:
```diff
@@ -210,43 +234,71 @@ void Index::passTsvFileIntoIdVector(const string& tsvFile, ExtVec& data,
+  if (items.size() > 0) {
+    // write remainder
+    
+    std::cout << "writing partial vocab no. " << numFiles << std::endl;
```
`LOG(LEVEL)`

niklas88 commented on src/index/Index.cpp:
```diff
@@ -210,43 +234,71 @@ void Index::passTsvFileIntoIdVector(const string& tsvFile, ExtVec& data,
   NTriplesParser p(ntFile);
-  google::sparse_hash_map<string, Id> vocabMap = _vocab.asMap();
+  LOG(INFO) << "Reading partial vocab ...\n";
+  google::sparse_hash_map<string, Id> vocabMap = vocabMapFromPartialIndexedFile(_onDiskBase + "partialVocabulary0");
```
I think we should stick with using a `.extension` scheme (i.e. add a `.`), also others are all lowercase with `-` word seperators

niklas88 commented on src/index/Index.cpp:
```diff
@@ -256,12 +308,32 @@ void Index::passNTriplesFileIntoIdVector(const string& ntFile, ExtVec& data,
+    if (i % linesPerPartial == 0) {
+      numFiles++;
       LOG(INFO) << "Lines processed: " << i << '\n';
+      std::cout << "reading partial vocab no. " << numFiles << std::endl;
```
`LOG(INFO)`

niklas88 commented on src/index/Index.h:
```diff
@@ -293,11 +293,22 @@ class Index {
+					  bool onDiskLiterals);
+
+  // ___________________________________________________________________
   size_t passNTriplesFileForVocabulary(const string& ntFile,
```
Add a comment describing what this function does, especially

niklas88 commented on src/index/Vocabulary.cpp:
```diff
@@ -51,6 +51,22 @@ void Vocabulary::writeToFile(const string& fileName) const {
+  std::ofstream out(fileName.c_str(), std::ios_base::out | std::ios_base::binary);
+  for (size_t i = 0; i < _words.size(); ++i) {
+    // 32 bits should be enough for len of string
+    unsigned int len = _words[i].size();
```
If it's always 32 bit we should use `uint32_t`

niklas88 commented on src/index/Vocabulary.cpp:
```diff
@@ -51,6 +51,22 @@ void Vocabulary::writeToFile(const string& fileName) const {
+  for (size_t i = 0; i < _words.size(); ++i) {
+    // 32 bits should be enough for len of string
+    unsigned int len = _words[i].size();
+    size_t zeros = 0;
```
This is an offset, right? If so use `off_t`

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+
+#include "../util/Log.h"
+
+class PairCompare {
```
This should probably get a more telling name

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  std::vector<std::fstream> infiles;
+  std::vector<std::vector<std::string>> vecs(numFiles);
+  std::ofstream outfile(basename + ".vocabulary");
+  std::ofstream outfileExternal(basename + "externalTextFile");
```
This should also use the common `".extension"` scheme

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  std::vector<std::vector<std::string>> vecs(numFiles);
+  std::ofstream outfile(basename + ".vocabulary");
+  std::ofstream outfileExternal(basename + "externalTextFile");
+  std::vector<bool> endOfFile(numFiles, false);
```
I think this name can be confused with the common meaning of `EOF` while it's really about having finished processing that file, or am I misunderstanding its use?

niklas88 commented on src/index/ExternalVocabulary.cpp:
```diff
@@ -3,6 +3,9 @@
 
 #include "./ExternalVocabulary.h"
+
+#include <fstream>
```
We use our own `util::File` class everywhere else I mean it clearly has it's faults but we should probably keep things consistent.

joka921 replied:
I would argue that for this cas (reading and writing text files line by line), std::getline on an fstream is better style and, according to a comment in the util::File class, also more performant. I would argue towards refactoring towards the std library when using textfiles in a line-by-line manner and util::File when reading and writing binary, because there some wrapping can not be wrong.

niklas88 replied:
Interesting idea, doing a quick search it looks like the only (non test) use of `File::readLine()` is in the [Text Index](https://github.com/ad-freiburg/QLever/blob/3667d45cc625610106ce82465ffbda5600745071/src/index/Index.Text.cpp#L42) reading docs files. So yeah I think it wouldn't be too much work getting rid of this use of `File`.

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  std::vector<std::streampos> posNextWord;
+  std::vector<std::vector<std::string>::iterator> iterators;
+  for (size_t i = 0; i < numFiles; i++) {
+    infiles.emplace_back(basename + "partialVocabulary" + std::to_string(i), std::ios_base::in | std::ios_base::out);
```
Extension naming scheme

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+void mergeVocabulary(const std::string& basename, size_t numFiles) {
+  size_t bufferSize = 200000;
+  std::vector<std::fstream> infiles;
+  std::vector<std::vector<std::string>> vecs(numFiles);
```
This deserves a more telling name like `mergeBuf` or something. Maybe also add a comment on how long the inner vectors will get, if I understand correctly each gets up to lines buffered.  Also maybe in the future if we turn that into a compile time constant we may be able to use `std::array` right? Would save an indirection.

joka921 replied:
Some thoughts on this:
1. Currently, nothing gets buffered at all, since this is way more complicated with the binary format which has the IDs in-place. I'm also not sure, if buffering makes sense in this case because we have to write the id of the word anyway and I think the paging of the file system should provide us some buffering or am I wrong in this case?
2. The number of partial files is not a compile-time constant, and if we use a reasonably big buffer this would be to big for a stack-places std::array.

Let me know what you think so I can refactor this function in a more readable way

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  std::ofstream outfileExternal(basename + "externalTextFile");
+  std::vector<bool> endOfFile(numFiles, false);
+  std::vector<std::streampos> posNextWord;
+  std::vector<std::vector<std::string>::iterator> iterators;
```
This could also be named more tellingly, these aren't just iterators, they are the current position for each file, right? So maybe `currPositions`

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  for (size_t i = 0; i < numFiles; i++) {
+    if (!vecs[i].empty()) {
+      queue.push(std::make_pair(vecs[i].front(), i));
+      //std::cout << vecs[i].front() << i << std::endl;;
```
remove unused debug statemens

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  using pair_T = std::pair<std::string, size_t>;
+  std::priority_queue<pair_T, std::vector<pair_T>, PairCompare> queue;
+  for (size_t i = 0; i < numFiles; i++) {
+    if (!vecs[i].empty()) {
```
what does it mean when this is empty?

joka921 replied:
I have added a comment, as stated above there must be a change here anyway.

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  std::string lastWritten = "";
+  size_t totalWritten = 0;
+
+  while (! queue.empty()) {
```
remove space

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+      //std::cout << "outputting " << top.first << " from " << top.second << std::endl;
+    } else {
+      // always write Index (also in case of duplicates)
+      // we already have increased total written, so for the duplicate
```
Have we? I only see the increment in the other side of the if? And I don't see a subtract one

niklas88 replied:
Ok fixed

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+    // refill with top element from current vector
+    iterators[top.second]++;
+    if (iterators[top.second] == vecs[top.second].end()) {
+      if (!endOfFile[top.second]) {
```
Then a `continue` on the negative might be easier to read, I'm not sure though

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+        auto i = top.second;
+        std::string word;
+        endOfFile[top.second] = true;
+        unsigned int len;
```
`uint32_t`

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+        if (infiles[i].read((char*)&len, sizeof(len))) {
+          vecs[i].emplace_back();
+          vecs[i].back().resize(len);
+          infiles[i].read(&(vecs[i].back()[0]), len);
```
I hope we can soon use C++17, non const `std::vector::data()` really is much nicer to read

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+google::sparse_hash_map<string, Id> vocabMapFromPartialIndexedFile(const string& partialFile) {
+  std::ifstream file(partialFile, std::ios_base::binary);
+  google::sparse_hash_map<string, Id> vocabMap;
+  unsigned int len;
```
`uint32_t`

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,125 @@
+  while (file.read((char*)&len, sizeof(len))) {
+    std::string word;
+    word.resize(len);
+    file.read(&(word[0]), len);
```
Remove parenthesis

joka921 replied:
[] has indeed higher precedence than & but I still find this clearer to read, but can remove it if you wish, since I now know about the preference

niklas88 replied:
no strong preference but looked weird to me

niklas88 commented on src/index/VocabularyGenerator.h:
```diff
@@ -0,0 +1,24 @@
+using std::string;
+// _______________________________________________________________
+// merge the partial vocabularies at basenamepartialVocabulary0 to basename +
+// partialVocabulary + numFiles-1
```
Why always with the last?

niklas88 commented on src/index/VocabularyGeneratorMain.cpp:
```diff
@@ -0,0 +1,5 @@
+#include "./VocabularyGenerator.h"
+
```
I thought this is integrated with the index building, or is this just for testing? That said there should be a `VocabularyGeneratorTest`

niklas88 commented:
@joka921 yes, that single commit version looks a lot cleaner. I think you have done well, this stuff is quite tricky. Just two years ago I had to learn that same procedure on a PR for the Go programming language standard library, only that reviewer was the founder of LiveJournal and his time reviewing my changes (and mistakes) probably cost Google a couple hundred bucks when he could have fixed them in far less time. Still I learned a lot and it's definitely come in handy already

niklas88 commented on src/index/ExternalVocabulary.cpp:
```diff
@@ -56,6 +59,27 @@ void ExternalVocabulary::buildFromVector(const vector<string>& v,
+// _____________________________________________________________________________
+void ExternalVocabulary::buildFromTextFile(const string& textFileName,
+                                         const string& outFileName) {
+  _file.open(outFileName.c_str(), "w");
```
This is missing a check if the `open()` actually worked. I just ran this in docker and had wrong permissions on the volume where the index is created and only when creating the real index did I see it fail.

joka921 replied:
Fixed

niklas88 commented:
One more thing, am I seeing this right and the partial vocabularies are currently not deleted after writing the full vocabulary? Ideally deleting them would also mean that we could continue after writing only some of them, however I think to work correctly we would need to first write each partial vocabulary and then atomically rename (mv) it to the final name.

niklas88 commented:
On my Wikipedia Freebase (not Easy) index build I'm currently getting a lot of `not found in partial Vocab` log entries, is this normal?
    
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: "9788420743523"@en
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0b1ldcq>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/common.topic.notable_types>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0775xx5>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0b1ldcq>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/key/soft.isbn>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: "9788420743523"
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0b1ldcq>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/key/soft.isbn>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: "9788420743523"
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0b1ldcq>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/common.topic.notable_for>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/g.1258_r_bd>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0b1ldcq>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/type.object.permission>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0775xvm>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/m.0b1ldcq>
    Fri Jun 15 09:25:12.107 - INFO:  not found in partial Vocab: <http://rdf.freebase.com/ns/type.object.key>

joka921 commented:
I have just fixed your problem. This is not normal and an error, I should let the program terminate at that point.
There was a wrong name for the first partialVocabulary file (error during refactoring). I fixed it and also added an assertion which checks this. When I have time I think I should write an end-to-end test for the index creation to avoid stuff like this

niklas88 commented on src/index/Index.cpp:
```diff
@@ -91,7 +91,7 @@ Index::ExtVec Index::createExtVecAndVocabFromNTriples(const string& ntFile,
   size_t nofLines = passNTriplesFileForVocabulary(ntFile, onDiskLiterals, NUM_TRIPLES_PER_PARTIAL_VOCAB);
   if (onDiskLiterals) {
-    _vocab.externalizeLiteralsFromTextFile(onDiskBase + "externalTextFile", onDiskBase + ".literals-index");
+    _vocab.externalizeLiteralsFromTextFile(onDiskBase + ".externalTextFile", onDiskBase + ".literals-index");
```
Rename this suffix, all other use `foo-bar` instead of `fooBar`, also `File` as a suffix adds no information so I propose `.externalized-text`

niklas88 commented on src/global/Constants.h:
```diff
@@ -51,3 +52,4 @@ static const int DEFAULT_NOF_DATE_YEAR_DIGITS = 19;
 // How many lines are parsed at once during index creation.
 // Reduce to save RAM
 static const int NUM_TRIPLES_PER_PARTIAL_VOCAB = 100000000;
+static const std::string PARTIAL_VOCAB_FILE_NAME = ".partial_vocabulary";
```
Use `-`instead of `_` to make even more consistent. I think I misstated it for snake case in the last review.

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -20,12 +21,15 @@ void mergeVocabulary(const std::string& basename, size_t numFiles) {
   std::ofstream outfile(basename + ".vocabulary");
-  std::ofstream outfileExternal(basename + "externalTextFile");
+  AD_CHECK(outfile.is_open());
+  std::ofstream outfileExternal(basename + ".externalTextFile");
```
Probably should make this a constant like for `.partial-vocabulary`

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -73,8 +77,8 @@ void mergeVocabulary(const std::string& basename, size_t numFiles) {
-      infiles[top.second].write((char*)&totalWritten, sizeof(totalWritten));
-
+      auto minusOne = totalWritten - 1;
+      infiles[top.second].write((char*)&minusOne, sizeof(totalWritten));
```
use `sizeof()` on the same variable to be more consistent.

niklas88 commented on test/CMakeLists.txt:
```diff
@@ -64,5 +68,6 @@ add_library(tests
\ No newline at end of file
+	    SparsehashTest
+	    VocabularyGeneratorTest
+            )
```
This `)` seems to be wrongly indented

niklas88 commented on test/VocabularyGeneratorTest.cpp:
```diff
@@ -0,0 +1,149 @@
+  // may throw?
+  MergeVocabularyTest() {
+    // name of random subdirectory
+    _basePath = boost::filesystem::unique_path("qleverVocTest%%%%_%%%%_%%%%_%%%%");
```
Still some Boost use or am I looking at outdated code?

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,139 @@
+  uint32_t len;
+  while (file.read((char*)&len, sizeof(len))) {
+    std::string word;
+    word.resize(len);
```
Instead of initializing a default string and only then resizing it probably makes more sense to create a string of the correct size filled with zeros using the size based constructor.

As for the buffering since we read sequentially and only know the size after reading the length I don't think custom buffering makes sense here. Let's just leave that to the OS. Alternatively it may also make more sense to read the whole `partialFile` into a single buffer in one go if it's not too large. That would save some syscalls.

niklas88 commented on src/index/Index.cpp:
```diff
@@ -159,6 +160,11 @@ void Index::createFromNTriplesFile(const string& ntFile,
                    _patterns, _maxNumPatterns);
   }
   openFileHandles();
+  // we have deleted the vocabulary in the process to save ram, so now we have
```
So since it is in `Index.cpp` it's not just for text indices?

joka921 replied:
You're right, this is a duplicate.
The actual solution is by setting _onDiskLiterals and _onDiskBase before, then it suffices to read it  in Index.Text.cpp. I am going to fix this tomorrow.

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -22,103 +22,80 @@ class PairCompare {
-      auto i = top.second;
+    auto i = top.second;
+    endOfFile[top.second] = true;
+    uint32_t len;
```
`size_t`

niklas88 commented on src/index/Index.cpp:
```diff
@@ -84,19 +85,48 @@ void Index::createFromTsvFile(const string& tsvFile, const string& onDiskBase,
+  if (!keepTempFiles) {
+    // remove temporary files only used during index creation
+    LOG(INFO) << "Removing temporary files (partial vocabulary and external text file...\n";
+    string removeCommand1 = "rm " + onDiskBase + EXTERNAL_LITS_TEXT_FILE_NAME;
```
use `"rm --" ` to make sure no one can pass e.g. `-rf` as part of `onDiskBase`. Also I'm not sure if we really should use the `system()` function here, sadly the only nice [standard way](https://en.cppreference.com/w/cpp/filesystem/remove) needs C++17. Now that Ubuntu 18.04 is out I'd be willing to use that in QLever though for a lot of our machines that means it really only works in a docker container at the moment. So maybe for now just add a TODO

niklas88 commented on src/index/Index.cpp:
```diff
@@ -84,19 +85,48 @@ void Index::createFromTsvFile(const string& tsvFile, const string& onDiskBase,
                                    const string& onDiskBase,
-                                   bool allPermutations, bool onDiskLiterals) {
+                                   bool allPermutations, bool onDiskLiterals,
+				   bool keepTempFiles) {
```
This indent looks wrong, did you try using our `.clang-format` settings?

niklas88 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -0,0 +1,116 @@
+  
+  std::vector<std::fstream> infiles;
+  // currently only one word is buffered, no matter how big the bufferSize is
+  // TODO: buffer, or throw the buffer out
```
If you threw it out, remove this comment

niklas88 commented on test/CMakeLists.txt:
```diff
@@ -52,9 +52,13 @@ target_link_libraries(SparsehashTest gtest_main -pthread)
 add_library(tests
             SparqlParserTest
-            StringUtilsTest
+	    StringUtilsTest
```
This look wrongly indented

niklas88 commented on test/CMakeLists.txt:
```diff
@@ -69,4 +73,5 @@ add_library(tests
             SparsehashTest
             GroupByTest
-            )
+	    VocabularyGeneratorTest
```
also mismatched indent

niklas88 commented on test/VocabularyGeneratorTest.cpp:
```diff
@@ -0,0 +1,175 @@
+    _path0 = std::string(PARTIAL_VOCAB_FILE_NAME + std::to_string(0));
+    _path1 =  std::string(PARTIAL_VOCAB_FILE_NAME + std::to_string(1));
+    // those names can be random
+    _pathExp0 = std::string(".partialVocabExp0");
```
I thought we moved to `snake_case_names`?

joka921 replied:
I think for filenammes it was names-with-dashes

joka921 commented on src/index/IndexBuilderMain.cpp:
```diff
@@ -217,14 +222,24 @@ int main(int argc, char** argv) {
-                              onDiskLiterals);
-    } else {
-      index.createFromOnDiskIndex(baseName, allPermutations, onDiskLiterals);
+    // TODO(j.kalmbach): onDiskLiterals is now  redundant in all other functions, probably
```
removed it everywhere but for  createFromOnDiskIndex, there the name  is so integral to the function that it is probably better  to use an argument.

joka921 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -22,100 +22,78 @@ class PairCompare {
       }
+
+      // write id to partial vocabulary
+      infiles[top.second].seekp(infiles[top.second].tellp());
```
I am not sure what you mean by "a seek() to the end". We are not at the end of the file but in the middle (the ids which we write are  directly next to the words we are reading.
This worked and there was some trouble with the flush(). I am also not sure about the performance, maybe this seek can be cheaper than a full flush to enforce the standard-compliance?
Could you in this case agree with explaining this by a comment?

joka921 replied:
Additionally, the interface of istream suggests, that the write and the read position are in fact two different things. They are implemented as one and the same position, but my version deals explicitly with the strangeness off the seekp and seekg function.

joka921 commented on src/index/VocabularyGenerator.cpp:
```diff
@@ -22,100 +22,78 @@ class PairCompare {
-  std::vector<std::vector<std::string>::iterator> iterators;
+
+  using pair_T = std::pair<std::string, size_t>;
+  std::priority_queue<pair_T, std::vector<pair_T>, PairCompare> queue;
```
I think this is not possible, since I need the third template parameter (the custom comparator) and one cannot "skip" template parameters to my knowledge.
